In [1]:
import pandas as pd

In [4]:
df = pd.read_csv('data/time_series_covid19_deaths_US.csv')

In [5]:
df.head()

UID iso2 iso3  code3    FIPS   Admin2 Province_State Country_Region  \
0  84001001   US  USA    840  1001.0  Autauga        Alabama             US   
1  84001003   US  USA    840  1003.0  Baldwin        Alabama             US   
2  84001005   US  USA    840  1005.0  Barbour        Alabama             US   
3  84001007   US  USA    840  1007.0     Bibb        Alabama             US   
4  84001009   US  USA    840  1009.0   Blount        Alabama             US   

         Lat      Long_  ... 11/16/20  11/17/20  11/18/20  11/19/20  11/20/20  \
0  32.539527 -86.644082  ...       36        37        37        39        39   
1  30.727750 -87.722071  ...       84        84        84        84        84   
2  31.868263 -85.387129  ...        9         9        10        10        10   
3  32.996421 -87.125115  ...       17        17        18        18        18   
4  33.982109 -86.567906  ...       34        34        35        35        35   

   11/21/20  11/22/20  11/23/20  11/24/20  11/25/20  
0        39        39        39        39        41  
1        84        84        84        84        98  
2        10        10        10        10        10  
3        17        17        17        17        17  
4        36        36        36        36        39  

[5 rows x 321 columns]

In [6]:
df.tail()

UID iso2 iso3  code3     FIPS      Admin2 Province_State  \
3335  84056039   US  USA    840  56039.0       Teton        Wyoming   
3336  84056041   US  USA    840  56041.0       Uinta        Wyoming   
3337  84090056   US  USA    840  90056.0  Unassigned        Wyoming   
3338  84056043   US  USA    840  56043.0    Washakie        Wyoming   
3339  84056045   US  USA    840  56045.0      Weston        Wyoming   

     Country_Region        Lat       Long_  ... 11/16/20  11/17/20  11/18/20  \
3335             US  43.935225 -110.589080  ...        2         2         2   
3336             US  41.287818 -110.547578  ...        4         4         4   
3337             US   0.000000    0.000000  ...        0         0         0   
3338             US  43.904516 -107.680187  ...        7         7         7   
3339             US  43.839612 -104.567488  ...        0         0         0   

      11/19/20  11/20/20  11/21/20  11/22/20  11/23/20  11/24/20  11/25/20  
3335         2         2         2         2         2         2         2  
3336         4         4         4         4         4         4         4  
3337         0         0         0         0         0         0         0  
3338         7         7         7         7         7         7         8  
3339         0         0         0         0         1         1         1  

[5 rows x 321 columns]

#### Some conclusions:

- 321 columns, 3340 entries
- No extra lines in start and end of file
- Many columns are dates, increamenting by one day at a time.
- Data end on Nov. 25 2020
- Contain geographical information: country - province - county + latitude & longitude
- We can see an entry (3337) with many zeroes



In [7]:
df.isnull()

UID   iso2   iso3  code3   FIPS  Admin2  Province_State  \
0     False  False  False  False  False   False           False   
1     False  False  False  False  False   False           False   
2     False  False  False  False  False   False           False   
3     False  False  False  False  False   False           False   
4     False  False  False  False  False   False           False   
...     ...    ...    ...    ...    ...     ...             ...   
3335  False  False  False  False  False   False           False   
3336  False  False  False  False  False   False           False   
3337  False  False  False  False  False   False           False   
3338  False  False  False  False  False   False           False   
3339  False  False  False  False  False   False           False   

      Country_Region    Lat  Long_  ...  11/16/20  11/17/20  11/18/20  \
0              False  False  False  ...     False     False     False   
1              False  False  False  ...     False     False     False   
2              False  False  False  ...     False     False     False   
3              False  False  False  ...     False     False     False   
4              False  False  False  ...     False     False     False   
...              ...    ...    ...  ...       ...       ...       ...   
3335           False  False  False  ...     False     False     False   
3336           False  False  False  ...     False     False     False   
3337           False  False  False  ...     False     False     False   
3338           False  False  False  ...     False     False     False   
3339           False  False  False  ...     False     False     False   

      11/19/20  11/20/20  11/21/20  11/22/20  11/23/20  11/24/20  11/25/20  
0        False     False     False     False     False     False     False  
1        False     False     False     False     False     False     False  
2        False     False     False     False     False     False     False  
3        False     False     False     False     False     False     False  
4        False     False     False     False     False     False     False  
...        ...       ...       ...       ...       ...       ...       ...  
3335     False     False     False     False     False     False     False  
3336     False     False     False     False     False     False     False  
3337     False     False     False     False     False     False     False  
3338     False     False     False     False     False     False     False  
3339     False     False     False     False     False     False     False  

[3340 rows x 321 columns]

In [8]:
null_in_column = df.isnull().sum()
null_in_column[null_in_column > 0]

FIPS      10
Admin2     6
dtype: int64

In [10]:
df[df.FIPS.isnull()]

UID iso2 iso3  code3  FIPS  \
1267  84070002   US  USA    840   NaN   
1304  84070005   US  USA    840   NaN   
1336  84070004   US  USA    840   NaN   
1591  84070003   US  USA    840   NaN   
2954  84070015   US  USA    840   NaN   
2959  84070016   US  USA    840   NaN   
2978  84070017   US  USA    840   NaN   
2979  84070018   US  USA    840   NaN   
2982  84070019   US  USA    840   NaN   
2990  84070020   US  USA    840   NaN   

                                         Admin2 Province_State Country_Region  \
1267                        Dukes and Nantucket  Massachusetts             US   
1304     Federal Correctional Institution (FCI)       Michigan             US   
1336  Michigan Department of Corrections (MDOC)       Michigan             US   
1591                                Kansas City       Missouri             US   
2954                                 Bear River           Utah             US   
2959                               Central Utah           Utah             US   
2978                             Southeast Utah           Utah             US   
2979                             Southwest Utah           Utah             US   
2982                                  TriCounty           Utah             US   
2990                               Weber-Morgan           Utah             US   

            Lat       Long_  ... 11/16/20  11/17/20  11/18/20  11/19/20  \
1267  41.406747  -70.687635  ...        2         2         2         2   
1304   0.000000    0.000000  ...        5         5         5         5   
1336   0.000000    0.000000  ...       76        76        77        77   
1591  39.099700  -94.578600  ...      244       249       249       250   
2954  41.521068 -113.083282  ...       24        24        25        25   
2959  39.372319 -111.575868  ...       14        15        15        15   
2978  38.996171 -110.701396  ...        6         6         6         6   
2979  37.854472 -111.441876  ...       70        72        73        77   
2982  40.124915 -109.517442  ...        4         4         4         4   
2990  41.271160 -111.914512  ...       42        43        44        46   

      11/20/20  11/21/20  11/22/20  11/23/20  11/24/20  11/25/20  
1267         2         2         2         2         2         2  
1304         5         5         5         5         5         5  
1336        79        80        80        80        80        81  
1591       253       255       255       255       269       276  
2954        25        25        25        25        26        28  
2959        15        16        16        16        16        18  
2978         6         7         7         7         7         7  
2979        79        80        81        83        83        86  
2982         4         4         4         4         5         6  
2990        46        49        50        50        52        52  

[10 rows x 321 columns]

In [12]:
df[df.Admin2.isnull()]

UID iso2 iso3  code3     FIPS Admin2            Province_State  \
100         16   AS  ASM     16     60.0    NaN            American Samoa   
336   84088888   US  USA    840  88888.0    NaN          Diamond Princess   
570   84099999   US  USA    840  99999.0    NaN            Grand Princess   
571        316   GU  GUM    316     66.0    NaN                      Guam   
2121       580   MP  MNP    580     69.0    NaN  Northern Mariana Islands   
3007       850   VI  VIR    850     78.0    NaN            Virgin Islands   

     Country_Region      Lat     Long_  ... 11/16/20  11/17/20  11/18/20  \
100              US -14.2710 -170.1320  ...        0         0         0   
336              US   0.0000    0.0000  ...        0         0         0   
570              US   0.0000    0.0000  ...        3         3         3   
571              US  13.4443  144.7937  ...       92        92        92   
2121             US  15.0979  145.6739  ...        2         2         2   
3007             US  18.3358  -64.8963  ...       23        23        23   

      11/19/20  11/20/20  11/21/20  11/22/20  11/23/20  11/24/20  11/25/20  
100          0         0         0         0         0         0         0  
336          0         0         0         0         0         0         0  
570          3         3         3         3         3         3         3  
571         92       103       103       103       103       103       109  
2121         2         2         2         2         2         2         2  
3007        23        23        23        23        23        23        23  

[6 rows x 321 columns]

substituting the null values

In [13]:
values_to_use = {'FIPS': -1, 'Admin2': 'Unassigned'}
df.fillna(value=values_to_use, inplace=True)

In [14]:
df[df.Admin2.isnull()]

Empty DataFrame
Columns: [UID, iso2, iso3, code3, FIPS, Admin2, Province_State, Country_Region, Lat, Long_, Combined_Key, Population, 1/22/20, 1/23/20, 1/24/20, 1/25/20, 1/26/20, 1/27/20, 1/28/20, 1/29/20, 1/30/20, 1/31/20, 2/1/20, 2/2/20, 2/3/20, 2/4/20, 2/5/20, 2/6/20, 2/7/20, 2/8/20, 2/9/20, 2/10/20, 2/11/20, 2/12/20, 2/13/20, 2/14/20, 2/15/20, 2/16/20, 2/17/20, 2/18/20, 2/19/20, 2/20/20, 2/21/20, 2/22/20, 2/23/20, 2/24/20, 2/25/20, 2/26/20, 2/27/20, 2/28/20, 2/29/20, 3/1/20, 3/2/20, 3/3/20, 3/4/20, 3/5/20, 3/6/20, 3/7/20, 3/8/20, 3/9/20, 3/10/20, 3/11/20, 3/12/20, 3/13/20, 3/14/20, 3/15/20, 3/16/20, 3/17/20, 3/18/20, 3/19/20, 3/20/20, 3/21/20, 3/22/20, 3/23/20, 3/24/20, 3/25/20, 3/26/20, 3/27/20, 3/28/20, 3/29/20, 3/30/20, 3/31/20, 4/1/20, 4/2/20, 4/3/20, 4/4/20, 4/5/20, 4/6/20, 4/7/20, 4/8/20, 4/9/20, 4/10/20, 4/11/20, 4/12/20, 4/13/20, 4/14/20, 4/15/20, 4/16/20, 4/17/20, 4/18/20, ...]
Index: []

[0 rows x 321 columns]

In [15]:
df.columns[:14]

Index(['UID', 'iso2', 'iso3', 'code3', 'FIPS', 'Admin2', 'Province_State',
       'Country_Region', 'Lat', 'Long_', 'Combined_Key', 'Population',
       '1/22/20', '1/23/20'],
      dtype='object')

In [20]:
df[df.columns[:11]].head()

UID iso2 iso3  code3    FIPS   Admin2 Province_State Country_Region  \
0  84001001   US  USA    840  1001.0  Autauga        Alabama             US   
1  84001003   US  USA    840  1003.0  Baldwin        Alabama             US   
2  84001005   US  USA    840  1005.0  Barbour        Alabama             US   
3  84001007   US  USA    840  1007.0     Bibb        Alabama             US   
4  84001009   US  USA    840  1009.0   Blount        Alabama             US   

         Lat      Long_          Combined_Key  
0  32.539527 -86.644082  Autauga, Alabama, US  
1  30.727750 -87.722071  Baldwin, Alabama, US  
2  31.868263 -85.387129  Barbour, Alabama, US  
3  32.996421 -87.125115     Bibb, Alabama, US  
4  33.982109 -86.567906   Blount, Alabama, US

Getting number of unique values in columns of interest

In [22]:
nunique = [(col, df[col].nunique()) for col in df.columns[:11]]
nunique

[('UID', 3340),
 ('iso2', 6),
 ('iso3', 6),
 ('code3', 6),
 ('FIPS', 3331),
 ('Admin2', 1978),
 ('Province_State', 58),
 ('Country_Region', 1),
 ('Lat', 3226),
 ('Long_', 3226),
 ('Combined_Key', 3340)]

Getting unique values in a specific column

In [27]:
iso2 = df.iso2.unique()
iso2

array(['US', 'AS', 'GU', 'MP', 'PR', 'VI'], dtype=object)

In [28]:
code3 = df['code3'].unique()
code3

array([840,  16, 316, 580, 630, 850])

In [24]:
iso3 = df.iso3.unique()
iso3

array(['USA', 'ASM', 'GUM', 'MNP', 'PRI', 'VIR'], dtype=object)

What does that mean?

ASM = American Samoa
GUM = Guam
MNP = Northern Mariana Islands
PRI = Puerto Rico
VIR = U.S. Virgin Islands

iso2, iso3 and code3 have the same number of unique entries.  
Checking if always the same combination is present or multiple combinations

In [29]:
to_combine = [iso2, iso3, code3]
to_combine

[array(['US', 'AS', 'GU', 'MP', 'PR', 'VI'], dtype=object),
 array(['USA', 'ASM', 'GUM', 'MNP', 'PRI', 'VIR'], dtype=object),
 array([840,  16, 316, 580, 630, 850])]

In [30]:
import itertools

In [31]:
all_possible_combinations = [el for el in itertools.product(*to_combine)]
len(all_possible_combinations)

216

In [32]:
all_possible_combinations[:5]

[('US', 'USA', 840),
 ('US', 'USA', 16),
 ('US', 'USA', 316),
 ('US', 'USA', 580),
 ('US', 'USA', 630)]

Combinations appearing in the data frame

In [33]:
actual_combinations = []

for el in all_possible_combinations:
    _ = df[(df.iso2 == el[0]) & 
           (df.iso3 == el[1]) & 
           (df.code3 == el[2])].shape[0]
    
    if _ > 0:
        actual_combinations.append((el, _))
        
actual_combinations

[(('US', 'USA', 840), 3256),
 (('AS', 'ASM', 16), 1),
 (('GU', 'GUM', 316), 1),
 (('MP', 'MNP', 580), 1),
 (('PR', 'PRI', 630), 80),
 (('VI', 'VIR', 850), 1)]

Two columns can be deleted without losing information

Column 'UID' not relevant to our analysis, column 'Combined_Key' is a combinations of the columns 'Admin2', 'Province_state', 'Country_region' and therefore also not needed

All four columns will be deleted

In [ ]:
df.drop(columns=['iso2', 'code3', 'UID', 'Combined_Key'], inplace=True)

In [37]:
df.columns[:11]

Index(['iso3', 'FIPS', 'Admin2', 'Province_State', 'Country_Region', 'Lat',
       'Long_', 'Population', '1/22/20', '1/23/20', '1/24/20'],
      dtype='object')

Examining entries with 0 longitude & latitude

In [38]:
df[df.Lat == 0]

iso3     FIPS      Admin2 Province_State Country_Region  Lat  Long_  \
52    USA  80001.0   Out of AL        Alabama             US  0.0    0.0   
64    USA  90001.0  Unassigned        Alabama             US  0.0    0.0   
89    USA  80002.0   Out of AK         Alaska             US  0.0    0.0   
95    USA  90002.0  Unassigned         Alaska             US  0.0    0.0   
111   USA  80004.0   Out of AZ        Arizona             US  0.0    0.0   
...   ...      ...         ...            ...            ...  ...    ...   
3233  USA  90054.0  Unassigned  West Virginia             US  0.0    0.0   
3285  USA  80055.0   Out of WI      Wisconsin             US  0.0    0.0   
3304  USA  90055.0  Unassigned      Wisconsin             US  0.0    0.0   
3329  USA  80056.0   Out of WY        Wyoming             US  0.0    0.0   
3337  USA  90056.0  Unassigned        Wyoming             US  0.0    0.0   

      Population  1/22/20  1/23/20  ...  11/16/20  11/17/20  11/18/20  \
52             0        0        0  ...         0         0         0   
64             0        0        0  ...         0         0         0   
89             0        0        0  ...         0         0         0   
95             0        0        0  ...         0         0         0   
111            0        0        0  ...         0         0         0   
...          ...      ...      ...  ...       ...       ...       ...   
3233           0        0        0  ...         0         0         0   
3285           0        0        0  ...         0         0         0   
3304           0        0        0  ...         0         0         0   
3329           0        0        0  ...         0         0         0   
3337           0        0        0  ...         0         0         0   

      11/19/20  11/20/20  11/21/20  11/22/20  11/23/20  11/24/20  11/25/20  
52           0         0         0         0         0         0         0  
64           0         0         0         0         0         0         0  
89           0         0         0         0         0         0         0  
95           0         0         0         0         0         0         0  
111          0         0         0         0         0         0         0  
...        ...       ...       ...       ...       ...       ...       ...  
3233         0         0         0         0         0         0         0  
3285         0         0         0         0         0         0         0  
3304         0         0         0         0         0         0         0  
3329         0         0         0         0         0         0         0  
3337         0         0         0         0         0         0         0  

[108 rows x 317 columns]

Checking for rows where all entries in dates are 0

In [39]:
df_dates = df[df.columns[8:]]
df_dates.head(2) 

1/22/20  1/23/20  1/24/20  1/25/20  1/26/20  1/27/20  1/28/20  1/29/20  \
0        0        0        0        0        0        0        0        0   
1        0        0        0        0        0        0        0        0   

   1/30/20  1/31/20  ...  11/16/20  11/17/20  11/18/20  11/19/20  11/20/20  \
0        0        0  ...        36        37        37        39        39   
1        0        0  ...        84        84        84        84        84   

   11/21/20  11/22/20  11/23/20  11/24/20  11/25/20  
0        39        39        39        39        41  
1        84        84        84        84        98  

[2 rows x 309 columns]

In [40]:
all_days_zero = [el for el in df_dates.sum(axis=1) if el==0]
len(all_days_zero)

319

Unfortunately I did not quite understand df.iloc[ ] df.loc[ ], df.at[ ] , df.iat[ ] or how this is useful to our analysis